shell_automaton/src/service/storage_service.rs

```diff
@@ -227,6 +228,55 @@ impl fmt::Debug for StorageServiceDefault {
     }
 }
 
+fn find_block_matching_context_storage_impl(
+    head_block_storage: BlockHeaderWithHash,
+    latest_context_hashes: &[ContextHash],
+    block_storage: &BlockStorage,
+) -> Result<Option<BlockHeaderWithHash>, StorageError> {
+    let mut head = head_block_storage;
+    let latest_context_hashes = HashSet::<&ContextHash>::from_iter(latest_context_hashes.iter());
+
+    while !latest_context_hashes.contains(head.header.context()) {
+        if head.header.level() == 0 {
+            return Ok(None);
+        }
+
+        head = match block_storage.get(head.header.predecessor())? {
+            Some(head) => head,
+            _ => return Ok(None),
+        };
+    }
+
+    Ok(Some(head))
+}
+
+fn find_block_matching_context_storage(
+    log: &slog::Logger,
+    head_block_storage: BlockHeaderWithHash,
+    latest_context_hashes: Vec<ContextHash>,
+    block_storage: &BlockStorage,
+) -> Result<BlockHeaderWithHash, StorageError> {
+    if latest_context_hashes.is_empty() {
+        return Ok(head_block_storage);
+    }
+
+    match find_block_matching_context_storage_impl(
+        head_block_storage.clone(),
+        &latest_context_hashes,
+        block_storage,
+    )? {
+        Some(head) => Ok(head),
+        None => {
+            slog::warn!(
```

Comment:
This should be an error, because the node will not be able to recover from this situation (other than by trying to find older context hashes to see if there is a match, or by re-starting the bootstrap process). But if this happens it is because something went very wrong and it is probably to just bail-out and notify the user of a potential corruption. What do you guys think?

Reply:
I was thinking that in the case where the context storage is level 1000, and the block storage is level 100, that `find_block_matching_context_storage` will not be able to find the corresponding block, since we only get the last few context hashes from the context.

> case where the context storage is level 1000, and the block storage is level 100

Not sure if it is a realistic scenario though